payments/models.py

```diff
@@ -154,3 +171,377 @@ def __str__(self):
+        related_name="lease",
+    )
+    _lease_object_id = models.UUIDField(null=True, blank=True, verbose_name=_("lease"))
+
+    objects = OrderManager()
+
+    def __str__(self):
+        return f"{self.product} [{self.status}]"
+
+    @property
+    @rounded
+    def pretax_price(self):
+        return convert_aftertax_to_pretax(self.price, self.tax_percentage)
+
+    @property
+    @rounded
+    def total_price(self):
+        return sum([ol.price for ol in self.order_lines.all()], self.price)
+
+    @property
+    @rounded
+    def total_pretax_price(self):
+        return sum(
+            [ol.pretax_price for ol in self.order_lines.all()], self.pretax_price
+        )
+
+    @property
+    @rounded
+    def total_tax_percentage(self):
+        return ((self.total_price - self.total_pretax_price) / self.total_price) * 100
+
+    def _check_valid_products(self) -> None:
+        if self.product and not isinstance(
+            self.product, (BerthProduct, WinterStorageProduct)
+        ):
+            raise ValidationError(_("You cannot assign other types of products"))
+
+    def _check_valid_lease(self) -> None:
+        # Check that only Berth or Winter leases are passed (not other models)
+        if not isinstance(self.lease, (BerthLease, WinterStorageLease)):
+            raise ValidationError(_("You cannot assign other types of leases"))
+
+        # Check that the lease customer and the received customer are the same
+        if self.lease.customer != self.customer:
+            raise ValidationError(
+                _("The lease provided belongs to a different customer")
+            )
+
+    def _check_product_and_lease(self) -> None:
+        if isinstance(self.product, BerthProduct) and not isinstance(
+            self.lease, BerthLease
+        ):
+            raise ValidationError(
+                _("A BerthProduct must be associated with a BerthLease")
+            )
+        elif isinstance(self.product, WinterStorageProduct) and not isinstance(
+            self.lease, WinterStorageLease
+        ):
+            raise ValidationError(
+                _("A WinterStorageProduct must be associated with a WinterStorageLease")
+            )
+
+    def _check_product_or_price(self) -> None:
+        if all([not self.product, not self._product_object_id, not self.price]):
+            raise ValidationError(
+                _("Order must have either product object or price value")
+            )
+
+    def _check_same_product(self, old_instance) -> None:
+        if all(
+            [old_instance.product, self.product, old_instance.product != self.product]
+        ):
+            raise ValidationError(_("Cannot change the product assigned to this order"))
+
+    def _check_same_lease(self, old_instance) -> None:
+        if old_instance.lease and self.lease != old_instance.lease:
+            raise ValidationError(
+                _("Cannot change the lease associated with this order")
+            )
```

Comment:
🙈  looking at this two and thinking more about it. it seems that the way `_check_same_lease` is safer, since with the check in `_check_same_product` you can now unlink some product from the order 🤔

Reply:
Hmm. Thinking about it, I think the `_check_same_product` is safer, because on `_check_same_lease` you could set the `lease` to `None` (since we're not checking for `self.lease`).

Reply:
exactly why `_check_same_lease` is safer :simple_smile: if an order had a lease, you can't set it to `None` anymore. with the product check you can.

Reply:
Also, on `_check_same_product`, `old_instance.product` would always be `True` since, product is `null=False`, so there will _always_ be a previous product. And _I think_ `self.product` would also always be `True` since the model constraint checks are run before, so if `None` is being passed, it would raise an error.

Reply:
> product is `null=False`
is it? 🤔

Reply:
AHHHHH NO. My bad, I was confused with `price`.

Also, now I got your point from it being safer. I think we should then also add a check for:
```python
if old_instance.lease and not self.lease:
    raise ValidationError(_("Cannot remove the lease associated with this order"))
```

Reply:
but 
```python
if old_instance.lease and self.lease != old_instance.lease:
            raise ValidationError(
                _("Cannot change the lease associated with this order")
            )
```
would already cover this case 🙂

Reply:
🤯 Always one step ahead of myself.